Cdd docs (#6305)
RasaHQ/rasa: docs/docs/conversation-driven-development.mdx

---
id: cdd

> sidebar_label: CDD

Author: Ja I was thinking the full name would be more helpful than the acronym

CDD includes the following actions:

*   **Share** your assistant with users as soon as possible
*   **Review** conversations on a regular basis
*   **Annotate** messages and use them as NLU training data
*   **Test** that your assistant always behaves as you expect
*   **Track** when your assistant fails and measure its performance over time
*   **Fix** how your assistant handles unsuccessful conversations

CDD is not a linear process; you’ll circle back to the same actions over and over as you develop and improve your bot.

Read more about these actions and the concept of CDD [here](https://blog.rasa.com/conversation-driven-development-a-better-approach-to-building-ai-assistants/). 

You can also check out [Rasa X](x), a purpose-built tool for CDD. 

## CDD in early stages of development

If you’re at the earliest stage of bot development, it might seem like CDD has no role to play - after all, you have no conversations yet! However, there are CDD actions you can take at the very beginning of bot development:

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Reviewer: I feel like 'best practices' sounds a bit too corporate to go in the docs

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Author: do you have an alternative in mind? this is referring to the whole sidebar section

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Reviewer: maybe something like "Learn more about ..." ?

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Author: sure, should the sidebar item still be called "Best Practices" though? Or should be rename that too?

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Reviewer: I'd advocating for taking that phrase out from the docs completely :D

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Author: "Guidelines for Bot Development"? @erohmensing

> 1. See the best practices for [NLU](http://s) and [Stories](http://s) for details on applying CDD in these areas.

Reviewer 2: Hm, it seemed like a pretty common term both on the forum and in interactions with enterprise users. But yeah, something like `Development Guidelines` could work?

2. Give your bot with test users early on. 

    CDD is all about listening to your users, so the earlier you find some, the better. 

    Test users can be anyone who doesn’t already know how your bot works from the inside. People on the bot development team should not be test users, since they know exactly what the bot can and can’t do. Don’t overinstruct your test users; they should have only as much knowledge of the bot’s domain as your end users will have.

> 3. Set up a CI/CD pipeline

Reviewer 2: ```suggestion
3. Set up a CI/CD pipeline.
```

>     CDD leads to frequent, smaller updates to your bot as you gather insights from bot conversations. [Setting up a CI/CD ](http://s)pipeline early on in development will enable you to act quickly on what you see in conversations

Reviewer 2: ```suggestion
    CDD leads to frequent, smaller updates to your bot as you gather insights from bot conversations. [Setting up a CI/CD pipeline](http://s)early on in development will enable you to act quickly on what you see in conversations.
```

At this stage, you can [install Rasa X in local mode](http://x) to make it easier to share your bot with test users, collect conversations, and apply NLU and Story best practices based on the conversations you collect.

## CDD with a bot in production 

> Once your bot is in production, you’ll have more conversations to gain insights from. 

Reviewer 2: ```suggestion
Once your bot is in production, you’ll have more conversations to gain insights from. Then you can fully apply CDD actions.
```

*   **Review**: Look in conversations for what users are really asking for.

    Your test users had at least some instruction about what the bot was intended to do; real users often either have no idea, or ignore instructions given to them. You can’t cater to every unexpected user behaviour, but you can try to address the main friction points you notice. Here are some things you could consider looking for

    *   Look at conversations where an “out_of_scope” intent or fallback behaviour occurred. These could indicate a potential new skill, or just a misclassified user utterance.
    *   Look for user frustration, such as requests for transfer to a human.

> *   **Annotate**: Continue to follow [best practices for NLU](http://s) as you add new user utterances from real conversations to your training data. Be careful not to overfit your NLU model to utterances like those already in your training data. This can happen when you continuously add user utterances that were already predicted correctly and with high confidence to your training data.  To avoid overfitting and help your model generalize to more diverse user utterances, add only user utterances that the model is predicting incorrectly or with low confidence. 

Reviewer 2: ```suggestion
*   **Annotate**: Continue to follow [best practices for NLU](http://s) as you add new user utterances from real conversations to your training data. Be careful not to overfit your NLU model to utterances like those already in your training data. This can happen when you continuously add user utterances that were already predicted correctly and with high confidence to your training data.  To avoid overfitting and help your model generalize to more diverse user utterances, add only user utterances that the model previously predicted incorrectly or with low confidence. 
```

> *   **Test:** Add successful user conversations to your test conversations. Doing this consistently will help ensure you don’t introduce regressions as you make other fixes to your bot. 

Reviewer 2: link on [test conversations] to "testing your assistant"

> *   **Track:** Look for clues to success and failure to help you track your bot’s performance

Reviewer 2: ```suggestion
*   **Track:** Look for clues to success and failure to help you track your bot’s performance.
```
